crates/transpiler/src/transpile_layout.rs

```diff
@@ -0,0 +1,493 @@
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use qiskit_circuit::nlayout::{NLayout, PhysicalQubit, VirtualQubit};
+
+/// The "layout" caused by transpilation
+///
+/// In general Qiskit's transpiler is unitary-preserving up to the initial layout
+/// and routing permutations. The initial layout permutation is caused by
+/// setting and applying the initial layout (the mapping from virtual circuit
+/// qubits to physical qubits on the target) and the routing permtuations are
+/// caused by swap gate insertion or permutation ellision prior to the initial
+/// layout. This struct tracks these details and provide an interface to reason
+/// about these permutations.
+pub struct TranspileLayout {
+    initial_layout: NLayout,
+    routing_permutation: Option<Vec<PhysicalQubit>>,
+    input_qubit_count: u32,
+    output_qubit_count: u32,
+}
+
+impl TranspileLayout {
+    pub fn new(
+        initial_layout: NLayout,
+        routing_permutation: Option<Vec<PhysicalQubit>>,
+        input_qubit_count: u32,
+        output_qubit_count: u32,
+    ) -> Self {
+        TranspileLayout {
+            initial_layout,
+            routing_permutation,
+            input_qubit_count,
+            output_qubit_count,
+        }
+    }
+
+    /// The number of input circuit qubits
+    pub fn num_input_qubits(&self) -> u32 {
+        self.input_qubit_count
+    }
+
+    /// The number of output circuit qubits
+    pub fn num_output_qubits(&self) -> u32 {
+        self.output_qubit_count
+    }
+
+    /// Generate an initial layout as an array of PhysicalQubit indices.
+    ///
+    /// # Args
+    ///     `filter_ancillas` - If set to `true` any ancilla qubits added to
+    ///     the circuit by the transpiler will not be included in the output
+    ///     array.
+    pub fn initial_layout(&self, filter_ancillas: bool) -> Vec<PhysicalQubit> {
+        if filter_ancillas {
+            (0..self.input_qubit_count)
+                .map(|x| {
+                    self.initial_layout
+                        .virtual_to_physical(VirtualQubit::new(x))
+                })
+                .collect()
+        } else {
+            (0..self.output_qubit_count)
+                .map(|x| {
+                    self.initial_layout
+                        .virtual_to_physical(VirtualQubit::new(x))
+                })
+                .collect()
+        }
+    }
+
+    /// Generate a routing permutation as an array of PhysicalQubit indices.
+    pub fn routing_permutation(&self) -> Vec<PhysicalQubit> {
+        match self.routing_permutation {
+            Some(ref perm) => perm.clone(),
+            None => (0..self.output_qubit_count)
+                .map(PhysicalQubit::new)
+                .collect(),
+        }
+    }
```

Comment:
If this is supposed to be a Rust-native struct, this would be better returning `Option<&[PhysicalQubit]>` rather than eagerly allocating.

Something like an `fn explicit_routing_permutation` could return `Cow<'self, [PhysicalQubit]>` and be essentially this method, but only allocate for the `None` case.